protocol/msgpack/unaggregated_iterator.go

```diff
@@ -0,0 +1,373 @@
+// It is not thread-safe.
+type unaggregatedIterator struct {
+	decoder           *msgpack.Decoder         // internal decoder that does the actual decoding
+	floatsPool        xpool.FloatsPool         // pool for float slices
+	policiesPool      pool.PoliciesPool        // pool for policies
+	metric            unaggregated.MetricUnion // current metric
+	versionedPolicies policy.VersionedPolicies // current policies
+	err               error                    // error encountered during decoding
+}
+
+// NewUnaggregatedIterator creates a new unaggregated iterator
+func NewUnaggregatedIterator(reader io.Reader, opts UnaggregatedIteratorOptions) (UnaggregatedIterator, error) {
+	if opts == nil {
+		opts = NewUnaggregatedIteratorOptions()
+	}
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+	it := &unaggregatedIterator{
+		decoder:      msgpack.NewDecoder(reader),
+		floatsPool:   opts.FloatsPool(),
+		policiesPool: opts.PoliciesPool(),
+	}
+
+	return it, nil
+}
+
+func (it *unaggregatedIterator) Reset(reader io.Reader) {
+	it.decoder.Reset(reader)
+	it.err = nil
+}
+
+func (it *unaggregatedIterator) Value() (*unaggregated.MetricUnion, policy.VersionedPolicies) {
+	return &it.metric, it.versionedPolicies
+}
+
+func (it *unaggregatedIterator) Err() error { return it.err }
+
+func (it *unaggregatedIterator) Next() bool {
+	if it.err != nil {
+		return false
+	}
+
+	// Resetting the metric to avoid holding onto the float64 slices
+	// in the metric field even though they may not be used
+	it.metric.Reset()
+
+	return it.decodeRootObject()
+}
+
+func (it *unaggregatedIterator) decodeRootObject() bool {
+	version := it.decodeVersion()
+	if it.err != nil {
+		return false
+	}
+	// If the actual version is higher than supported version, we skip
+	// the data for this metric and continue to the next
+	if version > unaggregatedVersion {
+		it.skip(it.decodeNumObjectFields())
+		return it.Next()
+	}
+	// Otherwise we proceed to decoding normally
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(rootObjectType)
+	if !ok {
+		return false
+	}
+	objType := it.decodeObjectType()
+	if it.err != nil {
+		return false
+	}
+	switch objType {
+	case counterWithPoliciesType, batchTimerWithPoliciesType, gaugeWithPoliciesType:
+		it.decodeMetricWithPolicies(objType)
+	default:
+		it.err = fmt.Errorf("unrecognized object type %v", objType)
+	}
+	it.skip(numActualFields - numExpectedFields)
+
+	return it.err == nil
+}
+
+func (it *unaggregatedIterator) decodeMetricWithPolicies(objType objectType) {
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(objType)
+	if !ok {
+		return
+	}
+	switch objType {
+	case counterWithPoliciesType:
+		it.decodeCounter()
+	case batchTimerWithPoliciesType:
+		it.decodeBatchTimer()
+	case gaugeWithPoliciesType:
+		it.decodeGauge()
+	default:
+		it.err = fmt.Errorf("unrecognized metric with policies type %v", objType)
```

Comment:
Seems like we do the type checks twice here, once in `decodeRootObject()` and another in the switch here.  Maybe we should just call `decodeMetricWithPolicies()` instead of doing the switch on line 109?